board_helper: support HDA link

brentlu opened the pull request:
1. Add HDA link support to board helper module.
2. Modify skl_hda_dsp_generic driver to create dai link array with board helper module.

brentlu committed:
ASoC: Intel: board_helpers: support HDA link initialization

Add a helper function for machine drivers to initialize HDA external
codec DAI link.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: Intel: skl_hda_dsp_generic: use common module for DAI links

Use intel_board module to create DAI link array for Intel iDisp HDMI,
HDA external codec, DMIC01, DMIC16K, and BT audio offload DAI BE links.

Signed-off-by: Brent Lu <brent.lu@intel.com>

plbossart commented:
@bardliao i'll let you review/merge

bardliao commented on sound/soc/intel/boards/sof_board_helpers.c:
```diff
@@ -624,6 +778,9 @@ sof_intel_board_get_ctx(struct device *dev, unsigned long board_quirk)
 	}
 
+	if (board_quirk & SOF_HDA_CODEC_PRESENT)
+		ctx->hda_codec_present = true;
```
Should we set `ctx->hda_codec_present` by HDA_EXT_CODEC(codec_mask)?

brentlu replied:
It's done in machine driver since we didn't pass codec_mask to board helper module.

bardliao replied:
> It's done in machine driver since we didn't pass codec_mask to board helper module.

Can we set `ctx->hda_codec_present` in machine driver? Just like [ctx->hdmi.idisp_codec = true;](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/skl_hda_dsp_generic.c#L232) I am thinking is SOF_HDA_CODEC_PRESENT really needed?

brentlu replied:
sure, we could remove that flag.

bardliao commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -100,10 +100,10 @@ static int skl_hda_audio_probe(struct platform_device *pdev)
+	card->name = "hda-dsp";
+	card->owner = THIS_MODULE;
+	card->fully_routed = true;
+	card->late_probe = skl_hda_card_late_probe;
```
Hmm, we need to add `Fixes: 33e59e50ee76 ("ASoC: Intel: skl_hda_dsp_generic: Allocate snd_soc_card dynamically")` in the commit message, and it is better to put the commit in the first commit. @ujfalusi FYI

brentlu replied:
done

ujfalusi replied:
omg, how did I done this??? and why it compiled in the first place...

ujfalusi replied:
@brentlu, can you add the Cc: stable@vger.kernel.org as well to make sure it is backported to stable.

brentlu replied:
Err...not sure how to do it...

ujfalusi replied:
You just need to add a line above the Fixes tag in the commit message:
```
Cc: stable@vger.kernel.org
```

brentlu replied:
done. Thanks~

brentlu committed:
ASoC: Intel: skl_hda_dsp_generic: fix clang/gcc build error

Use semicolon to complete an expression instead of comma for following
build error.

sound/soc/intel/boards/skl_hda_dsp_generic.c:108:2: error: expected expression
        dev_dbg(&pdev->dev, "board_quirk = %lx\n", board_quirk);

Cc: stable@vger.kernel.org
Fixes: 33e59e50ee76 ("ASoC: Intel: skl_hda_dsp_generic: Allocate snd_soc_card dynamically")
Signed-off-by: Brent Lu <brent.lu@intel.com>

bardliao commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -203,48 +48,80 @@ static void skl_set_hda_codec_autosuspend_delay(struct snd_soc_card *card)
+	unsigned long board_quirk = 0;
+	int ssp_bt;
+
+	if (mach_params->codec_mask & EXT_CODEC_MASK)
```
We can reuse `HDA_EXT_CODEC(mach_params->codec_mask)`

brentlu replied:
done

ujfalusi commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -278,6 +162,7 @@ module_platform_driver(skl_hda_audio)
 /* Module information */
 MODULE_DESCRIPTION("SKL/KBL/BXT/APL HDA Generic Machine driver");
 MODULE_AUTHOR("Rakesh Ughreja <rakesh.a.ughreja@intel.com>");
+MODULE_AUTHOR("Brent Lu <brent.lu@intel.com>");
```
Probably mention this in the comment message if it is really needed?

brentlu replied:
removed

ujfalusi commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -203,48 +48,80 @@ static void skl_set_hda_codec_autosuspend_delay(struct snd_soc_card *card)
+#define BT_OFFLOAD_BE_ID	8
+
+#define HDA_LINK_ORDER	SOF_LINK_ORDER(SOF_LINK_IDISP_HDMI,  \
+					SOF_LINK_HDA,        \
```
it is interesting that checkpatch is not complaining, but it would look better if the lines are aligned, or follow the other files and use 1 less TAB.

brentlu replied:
fixed

plbossart commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -278,6 +162,7 @@ module_platform_driver(skl_hda_audio)
 /* Module information */
 MODULE_DESCRIPTION("SKL/KBL/BXT/APL HDA Generic Machine driver");
 MODULE_AUTHOR("Rakesh Ughreja <rakesh.a.ughreja@intel.com>");
+MODULE_AUTHOR("Brent Lu <brent.lu@intel.com>");
```
My take is that references to 'author' should be avoided, be it in a file header or module description. If I added this to all the files I touched I would be the author of pretty much everything in SOF. same for Ranjani, Peter, Kai, Bard. The commit tamper-proof Signed-off-by is enough to give you credit for your work.

brentlu replied:
I just thought it is kinda mechanism that you could get notification when someone submit a patch to alsa-dev about it. I'll remove it from machine drivers if I got a chance.

ujfalusi commented on sound/soc/intel/boards/sof_board_helpers.c:
```diff
@@ -70,6 +70,66 @@ static int dmic_init(struct snd_soc_pcm_runtime *rtd)
+		return ret;
+	}
+
+	return 0;
```
You can simplify the code by:
```c
	if (ret)
		dev_err(rtd->dev, "fail to add hda routes, ret %d\n", ret);

	return ret;
```

brentlu replied:
done

bardliao commented:
SOFCI TEST
